session.go

```diff
@@ -907,12 +912,92 @@ func (s *session) DropPreparedStmt(stmtID uint32) error {
 	return nil
 }
 
+// StmtTxn wraps kv.Transaction to provide a new kv.Transaction.
+// It holds all statement related modification in the buffer before flush to the txn,
+// so if execute statement meets error, the txn won't be made dirty.
+type StmtTxn struct {
+	kv.Transaction
+
+	buf kv.MemBuffer
+	// TODO: Add binlog related data here.
+	// TODO: DirtyDB need to be statement level too.
```

Comment:
We can keep DirtyDB unmodified during the execution of a statement, only record the operations needs to be applied to the DirtyDB in StmtTxn, then apply them when flush.

Reply:
Yes, I'll do it right after this PR.